dag/image.go

```diff
@@ -37,11 +37,38 @@ type Image struct {
 	TestRunners     []types.TestRunner
 }
 
+type BuildReport struct {
+	ImageName                 string
+	BuildSuccess              bool
+	FailureMessage            string
+	DockerContextRelativePath string
+}
+
+func (img *Image) buildFailedReport(msg string) BuildReport {
```

Comment:
You don't need pointer receivers because you are not modifying any of `img` field.
We should always use value receiver because it is safer, especially when dealing with concurrency.
```suggestion
func (img Image) buildFailedReport(msg string) BuildReport {
```